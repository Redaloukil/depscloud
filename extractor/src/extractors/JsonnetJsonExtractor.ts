import {Dependency, DependencyManagementFile} from "@depscloud/api/v1alpha/deps";
import MatchConfig from "../matcher/MatchConfig";
import Extractor from "./Extractor";
import ExtractorFile from "./ExtractorFile";
import Languages from "./Languages";

const extract = (dependency: any, scope: string) => {
    return {
        name: parseName(dependency),
        organization:"",
        module,
        versionConstraint: dependency.version,
        scopes: [ scope ],
    };
}

const parseName = (dependency) => {
    return dependency.source.git.remote +"["+"#"+ dependency.source.git.subdir+"]";
}

    

class JsonnetJsonExtractor implements Extractor {
    matchConfig(): MatchConfig {
        return {
            includes: [
                "**/jsonnetfile.json",
            ],
            excludes: [],
        };
    }
    requires(): string[] {
        return [ "jsonnetfile.json" ];
    }
    public async extract(url: string, files: { [key: string]: ExtractorFile }): Promise<DependencyManagementFile> {
        const {
            version,
            dependencies,
            legacyImports,
        } = files["jsonnetfile.json"].json();

        const deps = dependencies.map(dependence => extract(dependence,""));


        return {
            language: Languages.JAVASCRIPT,
            system: "npm",
            sourceUrl: "",
            organization: "",
            module: "string",
            version,
            dependencies: deps,
            name: "",
        }
    }
}

export default  JsonnetJsonExtractor;