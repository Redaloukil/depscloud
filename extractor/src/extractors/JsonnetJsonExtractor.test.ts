import {readFile} from "fs";
import {promisify} from "util";
import ExtractorFile from "./ExtractorFile";
import JsonnetJsonExtractor from "./JsonnetJsonExtractor";

const readFileAsync = promisify(readFile);

describe("JsonnetJsonExtractor", () => {
    test("fullParse", async () => {
        const jsonPath = require.resolve("./testdata/jsonnetfile.json");
        const buffer = await readFileAsync(jsonPath);
        const content = buffer.toString();

        const parser = new JsonnetJsonExtractor();

        const actual = await parser.extract("", { "jsonnetfile.json": new ExtractorFile(content) });

        expect(actual).toMatchSnapshot();
    });
});
