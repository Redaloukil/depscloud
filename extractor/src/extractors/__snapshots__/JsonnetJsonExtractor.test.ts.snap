// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonnetJsonExtractor fullParse 1`] = `
Object {
  "dependencies": Array [
    Object {
      "module": "",
      "name": "https://github.com/grafana/grafonnet-lib.git[#grafonnet]",
      "organization": "",
      "scopes": Array [
        "dependencies",
      ],
      "versionConstraint": "master",
    },
    Object {
      "module": "",
      "name": "https://github.com/grafana/jsonnet-libs.git[#grafana-builder]",
      "organization": "",
      "scopes": Array [
        "dependencies",
      ],
      "versionConstraint": "master",
    },
  ],
  "language": "js",
  "module": "",
  "name": undefined,
  "organization": undefined,
  "sourceUrl": "",
  "system": "",
  "version": 1,
}
`;
